app/src/lib/stores/helpers/branch-pruner.ts

```diff
@@ -193,12 +193,25 @@ export class BranchPruner {
 
     // Create array of branches that can be pruned
     const candidateBranches = mergedBranches.filter(
-      mb => !ReservedRefs.includes(mb.canonicalRef)
+      ref => !ReservedRefs.includes(ref)
     )
 
-    const branchesReadyForPruning = candidateBranches.filter(
-      mb => !recentlyCheckedOutCanonicalRefs.has(mb.canonicalRef)
-    )
+    // get the locally cached branches of remotes (ie `remotes/origin/master`)
+    const remoteBranches = (await getBranches(
+      this.repository,
+      `refs/remotes/`
+    )).map(b => formatAsLocalRef(b.name))
+
+    const branchesReadyForPruning = candidateBranches.filter(ref => {
+      if (recentlyCheckedOutCanonicalRefs.has(ref)) {
+        return false
+      }
+      const upstream = getUpstreamRefForLocalBranchRef(ref, allBranches)
+      if (upstream === undefined) {
+        return false
```

Comment:
`getUpstreamRefForLocalBranchRef` looks like it has two conditions where it returns `undefined`:

 - unable to find the branch matching `ref` -> this seems a valid case to return `false` here
 - unable to find an upstream branch for `ref` -> should we return `true` here to ensure a branch without a local upstream branch is pruned here?

I worry we're conflating both of these cases and potentially not pruning some branches, but maybe I'm missing some context...

Reply:
in this context, unable to find an upstream ref means that we don't have upstream information for that branch, so the branch doesn't have an upstream (or for some reason we can't figure out what it is). in other words, its the just local ref for the upstream not the actual local branch (which may or may not actually exist).

line 213 is where we actually check to see if the upstream ref exists locally.

i've added some docs for `getUpstreamRefForLocalBranchRef` to clarify this logic. Does that sound right? (it took me more than a couple minutes to retrace my steps here in this logic, so please lmk if anything doesn't)

Reply:
Yeah, I'm still not clear on why it's fine to have the control flow like this. I wasn't clear on how else I could explain it, so I put together a test case that illustrates my thinking: #7860